lib/mixins/property-effects.js

```diff
@@ -409,14 +415,178 @@ function runReflectEffect(inst, property, props, oldProps, info) {
+      while ((info = queue.shift())) {
+        if (runComputedEffect(inst, '', changedProps, oldProps, info, hasPaths)) {
+          enqueueEffectsFor(info.methodInfo, computeEffects, queue, order, hasPaths);
+        }
+      }
       Object.assign(/** @type {!Object} */ (oldProps), inst.__dataOld);
       Object.assign(/** @type {!Object} */ (changedProps), inst.__dataPending);
-      inputProps = inst.__dataPending;
       inst.__dataPending = null;
+    } else {
+      // Original Polymer 2.x computed effects order, which continues running
+      // effects until no further computed properties have been invalidated
+      let inputProps = changedProps;
+      while (runEffects(inst, computeEffects, inputProps, oldProps, hasPaths)) {
+        Object.assign(/** @type {!Object} */ (oldProps), inst.__dataOld);
+        Object.assign(/** @type {!Object} */ (changedProps), inst.__dataPending);
+        inputProps = inst.__dataPending;
+        inst.__dataPending = null;
+      }
+    }
+  }
+}
+
+/**
+ * Inserts a computed effect into a queue, given the specified order. Performs
+ * the insert using a binary search.
+ * 
+ * Used by `orderedComputed: true` computed property algorithm.
+ * 
+ * @param {Object} info Property effects metadata
+ * @param {Array<Object>} queue Ordered queue of effects
+ * @param {Map<string,number>} order Map of computed property name->topological
+ *   sort order
+ */
+const insertEffect = (info, queue, order) => {
+  let start = 0;
+  let end = queue.length - 1;
+  let idx = -1;
+  while (start <= end) {
+    const mid = (start + end) >> 1;
+    // Note `methodInfo` is where the computed property name is stored in 
+    // the effect metadata
+    const cmp = order.get(queue[mid].methodInfo) - order.get(info.methodInfo);
+    if (cmp < 0) {
+      start = mid + 1;
+    } else if (cmp > 0) {
+      end = mid - 1;
+    } else {
+      idx = mid;
+      break;
+    }
+  }
+  if (idx < 0) {
+    idx = end + 1;
+  }
+  queue.splice(idx, 0, info);
+};
+
+/**
+ * Inserts all downstream computed effects invalidated by the specified property
+ * into the topologically-sorted queue of effects to be run.
+ * 
+ * Used by `orderedComputed: true` computed property algorithm.
+ * 
+ * @param {string} prop Property name
+ * @param {Object} computeEffects Computed effects for this element
+ * @param {Array<Obbject>} queue Topologically-sorted queue of computed effects
+ *   to be run
+ * @param {Map<string,numer>} order Map of computed property name->topological
+ *   sort order
+ * @param {boolean} hasPaths True with `changedProps` contains one or more paths
+ */
+const enqueueEffectsFor = (prop, computeEffects, queue, order, hasPaths) => {
+  const rootProperty = hasPaths ? root(prop) : prop;
+  const fxs = computeEffects[rootProperty];
+  if (fxs) {
+    for (let i=0; i<fxs.length; i++) {
+      const fx = fxs[i];
+      if ((fx.info.lastRun !== dedupeId) &&
+          (!hasPaths || pathMatchesTrigger(prop, fx.trigger))) {
```

Comment:
What does `hasPaths` have to do with this?

Reply:
It's just a perf optimization to avoid doing a more expensive check if `prop` is guaranteed to not be a path.  Carried over from [`runEffectForProperty`](https://github.com/Polymer/polymer/blob/b065d1450a4795d14633ece90510292229f6bc38/lib/mixins/property-effects.js#L161), which this method duplicates a bit of code from.